src/backends/policy_enforcement/souffle/value.h

```diff
@@ -0,0 +1,124 @@
+
+// The common supertype of any "fully fledged" Souffle value type. Anything
+// descending form `Value` should be able to be used in a `.type` declaration.
+// By this logic, `number`, `symbol`, any record, and any ADT are `Value`s, but
+// individual branches of an ADT are not.
+class Value {
+ public:
+  virtual std::string ToDatalogString() const = 0;
+};
+
+// This is the common supertype of any Souffle value type that can
+// participate in an ADT. This is all Values except Records. Values that can
+// participate in an ADT must have a virtual destructor.
+class AdtMemberValue : public Value {
+ public:
+  virtual ~AdtMemberValue() {}
+};
+
+// Corresponds to Souffle's `number` type.
+class Number : public AdtMemberValue {
+ public:
+  explicit Number(int64_t value) : number_value_(value) {}
+  virtual ~Number() {}
+
+  std::string ToDatalogString() const override {
+    return std::to_string(number_value_);
+  }
+
+ private:
+  int64_t number_value_;
+};
+
+// Corresponds to Souffle's `symbol` type.
+class Symbol : public AdtMemberValue {
+ public:
+  explicit Symbol(absl::string_view value) : symbol_value_(value) {}
+  virtual ~Symbol() {}
+
+  std::string ToDatalogString() const override {
+    return absl::StrFormat(R"("%s")", symbol_value_);
+  }
+
+ private:
+  std::string symbol_value_;
+};
+
+// A Record is, in fact, a value pointing to a tuple of values that
+// constitutes a record. This is necessary because Souffle conflates references
+// and the objects to which those references point. In the case that a type has
+// a recursive definition (common with linked lists), trying to implement this
+// without a pointer would cause layout issues.
+template <class... RecordFieldValueTypes>
+class Record : public Value {
+ public:
+  explicit Record() : record_arguments_() {}
+  explicit Record(RecordFieldValueTypes &&...args)
+      : record_arguments_(
+            std::make_unique<std::tuple<RecordFieldValueTypes...>>(
+                std::forward<RecordFieldValueTypes>(args)...)) {}
+
+  std::string ToDatalogString() const override {
+    if (!record_arguments_) return "nil";
+    return absl::StrFormat(
+        "[%s]", absl::StrJoin(*record_arguments_, ", ",
+                              [](std::string *out, const auto &arg) {
+                                absl::StrAppend(out, arg.ToDatalogString());
+                              }));
+  }
+
+ private:
+  std::unique_ptr<std::tuple<RecordFieldValueTypes...>> record_arguments_;
+};
+
+template <const char *kAdtName>
+class Adt : public AdtMemberValue {
+ public:
+  explicit Adt(absl::string_view branch_name) : branch_name_(branch_name) {}
+
+  std::string ToDatalogString() const {
+    return absl::StrFormat(
+        "$%s{%s}", branch_name_,
+        absl::StrJoin(members_, ", ", [](std::string *str, const auto &arg) {
+          absl::StrAppend(str, arg->ToDatalogString());
+        }));
+  }
+
+ protected:
+  std::vector<std::unique_ptr<AdtMemberValue>> members_;
```

Comment:
`arguments_` is more inline with ADT? i.e., arguments of a branch? Also, where or how is it populated?

Reply:
I can change this to `arguments_`.

I'm having the ADT branch (which is currently just a derived class of the ADT) populate the members in its constructor. That is less than ideal, as this means that there's shared ownership between `Adt` and the branch over this vector, but it does allow reducing the boilerplate you have to write when making one of these ADT branches a little (ie, you don't have to think about defining a field, just fill in the protected parent field that's already there).

Reply:
Changed to `arguments_`.